libraries/lib-mixer/EffectStage.cpp

```diff
@@ -14,32 +14,28 @@
 #include "EffectStage.h"
 #include "AudacityException.h"
 #include "AudioGraphBuffers.h"
-#include "SampleTrack.h"
+#include "WideSampleSequence.h"
 #include <cassert>
 
 namespace {
 std::vector<std::shared_ptr<EffectInstance>> MakeInstances(
    const EffectStage::Factory &factory,
-   EffectSettings &settings, double sampleRate, const SampleTrack &track
-   , std::optional<sampleCount> genLength, bool multi)
+   EffectSettings &settings, double sampleRate,
+   const WideSampleSequence &sequence,
+   std::optional<sampleCount> genLength, int channel)
 {
    std::vector<std::shared_ptr<EffectInstance>> instances;
    // Make as many instances as needed for the channels of the track, which
    // depends on how the instances report how many channels they accept
-   const auto range = multi
-      ? TrackList::Channels(&track)
-      : TrackList::Channels(&track).StartingWith(&track).EndingAfter(&track);
-   const auto nChannels = range.size();
-   size_t ii = 0;
-   for (auto iter = range.begin(); iter != range.end();) {
-      auto channel = *iter;
+   const auto nChannels = (channel < 0) ? sequence.NChannels() : 1;
```

Comment:
Looks like this function, when used by `PerTrackEffect`, takes care that `sequence` is always a leader. I wonder if this actually always must be true. If it must, please assert. I'm giving up on reviewing that part, at least for now, I can't grasp all possible situations.

Reply:
Just realized that `Mix.cpp` also uses this function, I reckon for the real-time effects. In this context `channel == -1`, but it's only called with leader sequences. So looks correct in that case too.

Reply:
> Side note :
> Now debugging splitting a stereo track to mono, as this is an "multichannel" effect.
> It hit `assert(!partner->mpGroupData);` in Track.cpp, line 216. Might be related to the work you did on having tracks share mute, gain and other properties. I guess this assert could be removed, but it will require your attention.

See commits 9f7d6c31b2d9fb2d4a9d01801673606830d9bab1 and c0f7aecfdb16c4cbd8cd240665a2fa7b5376d231, now merged.

Reply:
> Looks like this function, when used by `PerTrackEffect`, takes care that `sequence` is always a leader. I wonder if this actually always must be true. If it must, please assert. I'm giving up on reviewing that part, at least for now, I can't grasp all possible situations.

As I said, I can't assert that.  But as you will see in the next round of review, the condition must be ensured by `PerTrackEffect.cpp` to correct a bug in an earlier version of this branch, which caused assertion violations when generating a stereo chirp.